### PR TITLE
feat(ui): support popup service

### LINKED ISSUE #1566: [Quick Ticket] extract a canvas popup service to provide contains to upper features
### Initial checklist

- [X] Is this really a problem?
- [X] I have searched the Github Issues for similar issues, but did not find anything.

### Problem

A popup panel, anchored by a canvas element, is currently or will soon become a widely utilized component across various features. Some examples of its applications include data validation, filtering, and editing links. To meet this requirement efficiently and ensure a consistent user experience throughout different features, a dedicated UI service can be employed. This service would enable quick implementation and seamless integration of the popup panel functionality.

### Solution

Implement a service in package `@univerjs/ui` to provide such ability. Suggested names are `CanvasPopupService`  and `CanvasPopupPanelService`.

### PR DESCRIPTION
**Pull request by weird94**
close #1566

### CODE REVIEW

**Review by Jocs — APPROVED** [review]

**Comment by wzhudev** [comment]
Reviewing...

File: packages/design/src/components/popup/RectPopup.tsx [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+                    anchorRect,
+                    clientWidth,
+                    clientHeight,
+                    innerWidth,
+                    innerHeight
+                )
+            );
+        }
+    }, [
+        anchorRect.left,
```
Comment (wzhudev) [inline packages/design/src/components/popup/RectPopup.tsx:105]: Use comments to silent this warning.

File: packages/design/src/components/popup/RectPopup.tsx [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        const handleClick = (e: MouseEvent) => {
+            clickOtherFn.current?.(e);
+        };
+        setTimeout(() => {
+            window.addEventListener('click', handleClick);
+        }, 100);
+        return () => {
+            window.removeEventListener('click', handleClick);
+        };
+    }, [clickOtherFn]);
```
Comment (wzhudev) [inline packages/design/src/components/popup/RectPopup.tsx:122]: 这里的操作是为了避免重复绑定和解绑事件监听？这个 ref 看着有点怪。

File: packages/design/src/components/popup/RectPopup.tsx [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+     */
+    anchorRect: IAbsolutePosition;
+
+    direction?: 'horizontal' | 'vertical';
+
+    mask?: boolean;
+
+    onMaskClick?: React.MouseEventHandler<HTMLDivElement>;
+
+    onClickOther?: (e: MouseEvent) => void;
```
Comment (wzhudev) [inline packages/design/src/components/popup/RectPopup.tsx:41]: Could you please give furthur explanantion to this callback function? The name seems ambiguous to me.

File: packages/design/src/components/popup/RectPopup.tsx [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+            window.addEventListener('click', handleClick);
+        }, 100);
+        return () => {
+            window.removeEventListener('click', handleClick);
+        };
+    }, [clickOtherFn]);
+
+    return (
+        <>
+            {mask ? <div className={styles.popupMask} onClick={onMaskClick} /> : null}
```
Comment (wzhudev) [inline packages/design/src/components/popup/RectPopup.tsx:126]: Will it prevent clicking events be dispatced to the canvas?

File: packages/design/src/components/popup/RectPopup.tsx [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    direction?: 'horizontal' | 'vertical';
+
+    mask?: boolean;
+
+    onMaskClick?: React.MouseEventHandler<HTMLDivElement>;
+
+    onClickOther?: (e: MouseEvent) => void;
+}
+
+const calcHorizontalPopupPosition = (position: IAbsolutePosition, width: number, height: number, containerWidth: number, containerHeight: number): Partial<IAbsolutePosition> => {
```
Comment (wzhudev) [inline packages/design/src/components/popup/RectPopup.tsx:44]: These functions seems to be reusable in other components. We can extract them to a util dir.

File: packages/ui/src/views/components/popup/UniverPopup.tsx [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+            direction={popup.direction}
+            onClickOther={popup.onMaskClick}
+        >
+
+            {children}
+        </RectPopup>
+    );
+};
+
+export function UniverPopup() {
```
Comment (wzhudev) [inline packages/ui/src/views/components/popup/UniverPopup.tsx:50]: `CanvasPopup` could be a better name.

File: packages/ui/src/services/popup/global-popup-manager.service.ts [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    removeAll(): void;
+
+    popups$: Observable<[string, IPopup][]>;
+
+    get popups(): [string, IPopup][];
+}
+
+export const IGlobalPopupManagerService = createIdentifier<IGlobalPopupManagerService>('univer.popup.service');
+
+export class GlobalPopupManagerService implements IGlobalPopupManagerService {
```
Comment (wzhudev) [inline packages/ui/src/services/popup/global-popup-manager.service.ts:46]: CanvasPopupService

File: packages/ui/src/ui-plugin.ts [hunk 8]
```diff
@@ -18,6 +18,7 @@ import { ILocalStorageService, LocaleService, Plugin, PluginType } from '@univer
 import type { Dependency } from '@wendellhu/redi';
 import { Inject, Injector } from '@wendellhu/redi';
 
+import { GlobalPopupManagerService, IGlobalPopupManagerService } from '@univerjs/ui';
```
Comment (wzhudev) [inline packages/ui/src/ui-plugin.ts:21]: The import path is incorrect.

**Review by wzhudev — CHANGES_REQUESTED** [review]

**Review by wzhudev — COMMENTED** [review]
@jikkai Could you please help to review the design on the `RectPopup` component? I am concerned about the following:

1. Shall we implement a `ClickOutside` component?
2. Could `RectPopup` be implemented on the top of `Popup`?

File: packages/design/src/components/popup/RectPopup.tsx (hunk 4 above)
Comment (wzhudev) [reply packages/design/src/components/popup/RectPopup.tsx:126]: cc @jikkai `ClickOutside` component may optimize user experience here.

File: packages/design/src/components/popup/RectPopup.tsx (hunk 2 above)
Comment (weird94) [reply packages/design/src/components/popup/RectPopup.tsx:120]: 这个有点类似ahooks的 https://ahooks.js.org/zh-CN/hooks/use-memoized-fn

File: packages/design/src/components/popup/RectPopup.tsx (hunk 1 above)
Comment (weird94) [reply packages/design/src/components/popup/RectPopup.tsx:103]: dont

File: packages/design/src/components/popup/RectPopup.tsx (hunk 4 above)
Comment (weird94) [reply packages/design/src/components/popup/RectPopup.tsx]: `mask` has beed removed

File: packages/ui/src/views/components/popup/UniverPopup.tsx (hunk 6 above)
Comment (weird94) [reply packages/ui/src/views/components/popup/UniverPopup.tsx]: renamed to `GlobalPopup`, cause this popup can be use be for every element by given rect position and popup content, not only for canvas.

File: packages/ui/src/services/popup/global-popup-manager.service.ts (hunk 7 above)
Comment (weird94) [reply packages/ui/src/services/popup/global-popup-manager.service.ts:46]: This popup can be use be for every element by given rect position and popup content, not only for canvas.

File: packages/ui/src/ui-plugin.ts (hunk 8 above)
Comment (weird94) [reply packages/ui/src/ui-plugin.ts:21]: fixed

File: packages/design/src/components/popup/RectPopup.tsx (hunk 3 above)
Comment (weird94) [reply packages/design/src/components/popup/RectPopup.tsx]: Renamed to onClickOutside

File: packages/design/src/components/popup/RectPopup.tsx (hunk 5 above)
Comment (weird94) [reply packages/design/src/components/popup/RectPopup.tsx]: reusing these logic by using `RectPopup.calcPopupPosition`.

**Review by wzhudev — APPROVED** [review]
LGTM

**Pull request merged**